internals/overlord/state/notices.go

```diff
@@ -0,0 +1,412 @@
+	for _, notice := range s.notices {
+		if notice.id == id {
+			return notice
+		}
+	}
+	return nil
+}
+
+func (s *State) flattenNotices(filters NoticeFilters) []*Notice {
+	now := time.Now()
+	var notices []*Notice
+	for _, n := range s.notices {
+		if n.expired(now) || !filters.matches(n) {
+			continue
+		}
+		notices = append(notices, n)
+	}
+	return notices
+}
+
+func (s *State) unflattenNotices(flat []*Notice) {
+	now := time.Now()
+	s.notices = make(map[string]*Notice)
+	for _, n := range flat {
+		if n.expired(now) {
+			continue
+		}
+		uniqueKey := uniqueNoticeKey(n.noticeType, n.key)
+		s.notices[uniqueKey] = n
+	}
+}
+
+// WaitNotices waits for notices that match the filters to exist or occur,
+// returning the list of matching notices ordered by the last-repeated time.
+//
+// It waits till there is at least one matching notice or the context times
+// out or is cancelled. If there are existing notices that match the
+// filters, WaitNotices will return them immediately.
+func (s *State) WaitNotices(ctx context.Context, filters NoticeFilters) ([]*Notice, error) {
+	// State is already locked here by the caller, so notices won't be added
+	// concurrently.
+	notices := s.Notices(filters)
+	if len(notices) > 0 {
+		return notices, nil // if there are existing notices, return them right away
+	}
+
+	// Add a waiter channel for AddNotice to send to when matching notices arrive.
+	ch, waiterId := s.addNoticeWaiter(filters, ctx.Done())
+	defer s.removeNoticeWaiter(waiterId) // state will be re-locked when this is called
+
+	// Unlock state while waiting, to allow new notices to arrive (all state
+	// methods expect the caller to have locked the state before the call).
+	s.Unlock()
+	defer s.Lock()
+
+	select {
+	case notices = <-ch:
+		// One or more new notices arrived
+		return notices, nil
+	case <-ctx.Done(): // sender (processNoticeWaiters) also waits on this done channel
+		return nil, ctx.Err()
+	}
+}
+
+// noticeWaiter tracks a single WaitNotices call.
+type noticeWaiter struct {
+	filters NoticeFilters
+	ch      chan []*Notice
+	done    <-chan struct{}
+}
+
+// addNoticeWaiter adds a notice-waiter with the given filters. Processing
+// notices for this waiter stops when the done channel is closed.
+func (s *State) addNoticeWaiter(filters NoticeFilters, done <-chan struct{}) (ch chan []*Notice, waiterId int) {
+	s.noticeWaiterId++
+	waiter := noticeWaiter{
+		filters: filters,
+		ch:      make(chan []*Notice),
+		done:    done,
+	}
+	s.noticeWaiters[s.noticeWaiterId] = waiter
+	return waiter.ch, s.noticeWaiterId
+}
+
+// removeNoticeWaiter removes the notice-waiter with the given ID.
+func (s *State) removeNoticeWaiter(waiterId int) {
+	delete(s.noticeWaiters, waiterId)
+}
+
+// processNoticeWaiters loops through the list of notice-waiters, and wakes up
+// and sends to any that match the filters.
+func (s *State) processNoticeWaiters() {
+	for waiterId, waiter := range s.noticeWaiters {
+		notices := s.Notices(waiter.filters)
```

Comment:
I actually wrote it using `sync.Cond` initially, and it was a bit simpler, but unfortunately `Cond` with its `Wait()` method doesn't support cancellation/timeout. You can fake it by firing up a goroutine and pushing to a channel, but then you'll leak the goroutine if the timeout occurs before the event. So I went with a list of channels so that I can select on them.

Lots more reading [here](https://github.com/golang/go/issues/16620), where Brad Fitzpatrick suggests that Go adds a channel version of `Cond.Wait` to allow cancellation / selecting on it. But that hasn't been done or accepted yet.

Reply:
Please have a look here: https://pkg.go.dev/context#AfterFunc

This seems to solve the problem by allowing cancellation of the context to broadcast on the Cond, so the function can simply Wait on the Cond, and when a notice arrives it calls Broadcast for the waiters to observe as desired.

Probably not as a coincidence, the example there turns out to be about Cond too.

Reply:
Interesting! I hadn't seen `context.AfterFunc` before, and I hadn't thought of making the context cancellation do a broadcast. Unfortunately we can't use `AfterFunc` directly as it just arrived in Go 1.21 (we're on 1.20), but it's fairly easy to emulate what we need of it with a few lines of code.

Anyway, I've updated the PR to use `sync.Cond` ([diff here](https://github.com/canonical/pebble/pull/295/commits/e29f3f774a501816c859070b009bd596ddcfc95b)). It's definitely simpler and avoids the tedious data structures and channel management, and will simplify further once we can use `context.AfterFunc` directly.